File: AtCoder/JSC2019qual/C.py
N = int(input())
S = input()
from math import factorial

cnt = 0

ans = 1
mod = 10 ** 9 + 7
for i in range(2 * N):
    s = S[i]
    if s == "W":
        if cnt == 0:
            print(0)
            exit()
        else:
            if cnt % 2 == 0:
                ans *= cnt
                ans %= mod
                cnt -= 1
            else:
                cnt += 1
    else:
        if cnt % 2 == 0:
            cnt += 1
        else:
            ans *= cnt
            ans %= mod
            cnt -= 1
if cnt != 0:
    print(0)
    exit()
ans *= factorial(N)
ans %= mod
print(ans)

File: AtCoder/JSC2019qual/C.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int, int> P;
typedef long long ll;
#define rep(i, n) for(int i=0;i<(n);i++)
#define rep2(i, m, n) for(int i=m;i<(n);i++)
#define rrep(i, n, m) for(int i=n;i>=(m);i--)
using Graph = vector<vector<int>>;
const int dx[4] = {1, 0, -1, 0};
const int dy[4] = {0, 1, 0, -1};
const ll MOD = 1000000007;
#ifdef __DEBUG

#include "cpp-pyprint/pyprint.h"

#endif

void Main() {
    int N;
    string S;
    cin >> N >> S;
    ll cnt = 0;
    ll ans = 1;

    for (int i = 0; i < 2 * N; ++i) {
        if (S[i] == 'W') {
            if (cnt == 0) {
                cout << 0 << endl;
                return;
            }
            if(cnt%2 == 0){
                ans *= cnt;
                ans %= MOD;
                cnt --;
            }else{
                cnt ++;
            }
        } else {
            if (cnt % 2 == 0){
                cnt ++;
            }else{
                ans *= cnt;
                ans %= MOD;
                cnt --;
            }
        }
    }

    if (cnt != 0) {
        cout << 0 << endl;
    } else {
        for (ll i = 1; i <= N; ++i) {
            ans *= i;
            ans %= MOD;
        }
        cout << ans << endl;
    }

}

int main() {
    cin.tie(0);
    ios::sync_with_stdio(false);
    cout << fixed << setprecision(15);
    Main();
}
